turbopack/crates/turbopack-ecmascript-runtime/js/src/browser/runtime/base/build-base.ts

```diff
@@ -0,0 +1,155 @@
+
+/**
+ * Gets or instantiates a runtime module.
+ */
+// @ts-ignore
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+function getOrInstantiateRuntimeModule(
+  moduleId: ModuleId,
+  chunkPath: ChunkPath,
+): BaseModule {
+  const module = moduleCache[moduleId];
+  if (module) {
+    if (module.error) {
+      throw module.error;
+    }
+    return module;
+  }
+
+  return instantiateModule(moduleId, { type: SourceType.Runtime, chunkPath });
+}
+
+/**
+ * Retrieves a module from the cache, or instantiate it if it is not cached.
+ */
+// Used by the backend
+// @ts-ignore
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const getOrInstantiateModuleFromParent: GetOrInstantiateModuleFromParent<BaseModule> = (
+  id,
+  sourceModule
+) => {
+  const module = moduleCache[id];
+
+  if (sourceModule.children.indexOf(id) === -1) {
+    sourceModule.children.push(id);
+  }
+
+  if (module) {
+    if (module.parents.indexOf(sourceModule.id) === -1) {
+      module.parents.push(sourceModule.id);
+    }
+
+    return module;
+  }
+
+  return instantiateModule(id, {
+    type: SourceType.Parent,
+    parentId: sourceModule.id,
+  });
+};
+
+function instantiateModule(id: ModuleId, source: SourceInfo): BaseModule {
+  const moduleFactory = moduleFactories[id];
+  if (typeof moduleFactory !== "function") {
+    // This can happen if modules incorrectly handle HMR disposes/updates,
+    // e.g. when they keep a `setTimeout` around which still executes old code
+    // and contains e.g. a `require("something")` call.
+    let instantiationReason;
+    switch (source.type) {
+      case SourceType.Runtime:
+        instantiationReason = `as a runtime entry of chunk ${source.chunkPath}`;
+        break;
+      case SourceType.Parent:
+        instantiationReason = `because it was required from module ${source.parentId}`;
+        break;
+      case SourceType.Update:
+        instantiationReason = "because of an HMR update";
+        break;
+      default:
+        invariant(source, (source) => `Unknown source type: ${source?.type}`);
+    }
+    throw new Error(
+      `Module ${id} was instantiated ${instantiationReason}, but the module factory is not available. It might have been deleted in an HMR update.`
+    );
+  }
+
+  let parents: ModuleId[];
+  switch (source.type) {
+    case SourceType.Runtime:
+      runtimeModules.add(id);
+      parents = [];
+      break;
+    case SourceType.Parent:
+      // No need to add this module as a child of the parent module here, this
+      // has already been taken care of in `getOrInstantiateModuleFromParent`.
+      parents = [source.parentId];
+      break;
+    case SourceType.Update:
+      parents = source.parents || [];
+      break;
+    default:
+      invariant(source, (source) => `Unknown source type: ${source?.type}`);
+  }
+
+  const module: BaseModule = {
+    exports: {},
+    error: undefined,
+    loaded: false,
+    id,
+    parents,
+    children: [],
```

Comment:
parent and children could be omitted in production. They are primarily used for HMR and webpack also omits them in production